---
title: "Problem Set 3. Due Thurs March 2 5pm"
author: "Frank Santoriello"
date: "Feb 23, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>")
```

# Submission

1. Fork this repository to your own account
1. Make changes to the `README.Rmd` file (including the author field above).
1. Knit the file to HTML format and add it to your git repository (`git add`)
1. Submit a pull request with your Rmd and knitted HTML files.

# Overview 

You will examine a [time-course of mRNA expression][1] of barley powdery mildew fungus in immunocompromised plants (*Arabidopsis*). The fungus Blumeria graminis f. sp. hordei (called *Bgh* from now on)

Counts were generated using the [`maSigPro` software][3] and deposited in [NCBI GEO][2].

[1]: http://www.ncbi.nlm.nih.gov/pubmed/23175786
[2]: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43163  
[3]: http://bioinformatics.oxfordjournals.org/content/30/18/2598.full

Some important identifiers for the data set:

| Name | Description |
| ---- | ----------- |
| `A6` | *Bgh* isolate expressing other AVRA effectors |
| `K1` | *Bgh* isolate expressing the cognate AVRA1 effector for MLA1|
| `pps` | *Arabidopsis* plants |
| `B12` | *Arabidopsis* plants expressing MLA1-HA|

We will download tables from GEO containing counts of mRNA abundance from both the fungal sample.

# Raw data

First we need load the data.

```{r load_raw_data, message = FALSE}
library(tidyverse)
library(stringr)
library(gridExtra)

# bgh data
bgh_url <- 'http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE43163&format=file&file=GSE43163_CompleteCountTable_Bgh.txt.gz'

# 2 info and blank lines at top, skip them
raw_data <- read_tsv(bgh_url, skip = 2)

# the header for the first column is "NA", set it manually
names(raw_data)[1] <- 'gene.name'

raw_data <- as_data_frame(raw_data)
```

# Problems


## Problem 1

Now that the raw data are loaded, your first task is to tidy the data with `tidyr`. The `cleaned_data` should look something like this:

```
# A tibble: 310,896 Ã— 6
             gene.name plant fungus time.point   rep value
*                <chr> <chr>  <chr>      <chr> <chr> <int>
1             bgh04079   B12     A6       6hpi     1    13
2             bgh01634   B12     A6       6hpi     1    31
3  bghG000012000001001   B12     A6       6hpi     1   121
4  bghG000012000002001   B12     A6       6hpi     1     3
5             bgh00757   B12     A6       6hpi     1   253
6             bgh01273   B12     A6       6hpi     1    45
7             bgh01274   B12     A6       6hpi     1    32
8             bgh01277   B12     A6       6hpi     1     9
9             bgh06140   B12     A6       6hpi     1    47
10            bgh05774   B12     A6       6hpi     1    42
# ... with 310,886 more rows
```

The key steps are to `gather()` the data into key-value pairs and then `separate()` the information from the key column into new columns.

```{r problem_1}
# tidy data
raw_data %>%
  gather(key, value, -gene.name) %>%
  separate(key, into = c('plant','fungus','time.point','rep'), sep = '_') -> cleaned_data

# report the cleaned_data by just naming it, uncomment the following line:
cleaned_data
```

## Problem 2

You need to translate what you learned on the command line into the R world.

Which plant has the highest expression of any gene in the 6hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "6hpi"' | sort -k6nr | head -n 1 | cut -f2
```

```{r}
# translate to dplyr code
cleaned_data %>%
  filter(time.point == '6hpi') %>%
  arrange(desc(value)) %>%
  slice(1) %>%
  select(2) 

```

Which plant / fungus pair has the highest expression in the 18hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "18hpi"' | sort -k6nr | head -n 1 | cut -f2,3
```

```{r}
# translate to dplyr code
cleaned_data %>%
  filter(time.point == '18hpi') %>%
  arrange(desc(value)) %>%
  slice(1) %>%
  select(c(2,3))
```

## Problem 3

Identify the top 3 most consistently differentially expressed genes between the earliest and latest time points for each combination of `plant` and `fungus` strains.

- "Differential expression"" is the difference between `value` (i.e., gene expression level) between time points.

- "Consistency" is the smallest variance in `value` between replicates.

### Strategy

1. Create a new table from the cleaned data by moving each `hpi` value to a new column name, with counts for each in the column (hint: use a `tidyr` verb). It is helpful to reformat the `hpi` values by converting from e.g. `6hpi` to `hpi.6`. You can use `mutate` to do this, i.e.:

```{r reformat_time, eval = TRUE}
# reformat the data for differential expression analysis 
cleaned_data %>%
  mutate(time.point = str_replace(time.point, 'hpi', ''),
         time.point = str_c('hpi.', time.point)) %>%
  spread(time.point, value) -> expr_data

expr_data <- expr_data[,c(1:4,8,5:7)]

```

2. Create a new column containing the expression difference between the relevant time points.

3. Calculate summary statistics (mean and variance) of the expression differences by grouping (hint) the `gene.name`, `plant`, and `virus` columns.

4. Sort by these statistics and use the `dplyr` verb `slice` to pull the ones you want (i.e., the top 3). Note you will have to remove `gene.name` from the grouping so that sorting works.

```{r problem_3 }
# identify the top 3 consistently differentially expressed genes
expr_data %>%
  # log2 transform data and calculate fold change between hpi.6 and hpi.24
  mutate(hpi.6 = log(hpi.6, 2),
         hpi.12 = log(hpi.12, 2),
         hpi.18 = log(hpi.18, 2),
         hpi.24 = log(hpi.24, 2),
         expr.diff = hpi.24 - hpi.6) %>%
  group_by(gene.name, plant, fungus) %>%
  # calculate mean fold change and variance
  summarise(expr.mean = mean(expr.diff), expr.var = var(expr.diff)) %>%
  group_by(plant, fungus) %>%
  # remove infinite values and idenify genes with > 2 fold expression difference
  filter(abs(expr.mean) >= 1, 
         expr.mean != Inf, 
         expr.mean != -Inf) %>%
  # sort by varience to identify most consistently differential genes
  arrange(expr.var) %>%
  # identify 3 most consitently differential genes for each plant/fungus pair
  slice(1:3) -> diff_expr_data

diff_expr_data
  
```

## Problem 4

Now examine the above final data frame above and write a few sentences putting inline code in least 3 places. For example, There are `r nrow(mtcars)` rows of data in `mtcars`.

The tidying process for the data above generated a tibble with `r ncol(cleaned_data)` columns and `r nrow(cleaned_data)` rows. 

That data was then transformed for expression analysis by splitting out the `r colnames(cleaned_data[,4])` column into 4 new columns (`r colnames(expr_data[,5:8])`) corresponding to the different time points. These were then log transformed and analyzed for consistent differential expression resulting in the `r nrow(diff_expr_data)` genes listed in `diff_expr_data`.

## Problem 5

Plot the expression (`value`) by time (`hpi`) for the above genes. Format the plot as follows:

1. Plot each point.
1. Connect the points with a smooth line 
1. Create a facet for each `plant` / `bgh` pair 
1. Assign a different color to each gene 
1. Assign a different shape to each replicate 
1. Add a descriptive title to the plot 
1. Use the "black & white" theme

Extra credit: add error bars to the plot (use `geom_errorbar`).

```{r problem_5, fig.width = 10, fig.height = 10 }

cleaned_data %>%
  # reformat time.point
  mutate(time.point = str_replace(time.point, 'hpi', ''),
         time.point = as.numeric(time.point)) %>%
  # group by time.point and pull out genes that match the 12 genes listed in problem 3
  group_by(time.point) %>%
  subset(cleaned_data$gene.name %in% diff_expr_data$gene.name) %>%
  # plot value vs time.point
  ggplot(aes(x = factor(time.point), y = value, 
             color = gene.name, shape = rep, group = gene.name)) +
    geom_point() +
    geom_smooth(method = "glm", se = FALSE) +
    facet_wrap(c("plant", "gene.name"), scales = "free", ncol = 4) +
    theme_bw() +
    ggtitle("Consistently Differentially Expressed Genes by Plant Species")
    
```

Not all plant gene pairs are relavant based on the output of problem 3. I attempted some fancy subsetting to account for this, but failed. Therefore, here is all twelve genes for each plant. 
